File: src/com/train/algorithm/tree/implementInC++/SerializeandDeserializeBST.cpp
//
// Created by 罗斯 on 7/11/2022.
//
#include <string>
#include <iostream>

using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string s;
        serialize(root, s);
        return s;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return deserialize(data, pos, INT_MIN, INT_MAX);
    }
private:
    void serialize(TreeNode* root, string& s){
        if (!root) return;
        s.append(reinterpret_cast<const char *>(&root->val), sizeof(root->val));
        serialize(root->left, s);
        serialize(root->right, s);
    }
    TreeNode* deserialize(const string& s, int& pos, int curMin, int curMax){
        if (pos >= s.size()) return nullptr;
        int val = *reinterpret_cast<const int*>(s.data() + pos);
        if (val < curMin || val > curMax) return nullptr;
        pos += sizeof(val);
        TreeNode* root = new TreeNode(val);
        root->left = deserialize(s, pos, curMin, val);
        root->right = deserialize(s, pos, val, curMax);
        return root;
    }
};

File: src/com/train/algorithm/tree/implementInPython/SerializeAndDeserializeBST.py
from collections import deque


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        if not root: return ""
        stack = [root]
        serialized = ""
        while stack:
            node = stack.pop()
            if not node:
                serialized += "$,"
            else:
                serialized += str(node.val) + ","
                stack.append(node.right)
                stack.append(node.left)
        return serialized[:-1]

    def new_deserialize(self, queue):
        value = queue.popleft()
        if value == "$":
            return None
        node = TreeNode(int(value))
        node.left = self.new_deserialize(queue)
        node.right = self.new_deserialize(queue)
        return node

    def deserialize(self, data):
        if not data: return None
        values = data.split(",")
        queue = deque(values)
        return self.new_deserialize(queue)